### ISSUE TITLE
os_router OpenStack module does not work with TryStack

### ISSUE THREAD

**Issue by ibratoev** [ansible/ansible-modules-core#2474]
#### Issue Type:

Bug Report
#### Ansible Version:

ansible 2.0.0 (devel 634e10e5b8) last updated 2015/11/13 22:04:10 (GMT +300)
  lib/ansible/modules/core: (detached HEAD 6e37f1dcef) last updated 2015/11/13 22:07:25 (GMT +300)
  lib/ansible/modules/extras: (detached HEAD e52e015791) last updated 2015/11/13 22:07:25 (GMT +300)
#### Ansible Configuration:

default settings
#### Environment:

OSX
#### Summary:

os_router module does not work well with TryStack.
#### Steps To Reproduce:

Create a TryStack account and fill up properly your clouds.yml. Then:

``` yaml
    - name: Create OS network
      os_network:
        state: present
        name: testNetwork
        external: no
        cloud: trystack

    - name: Create OS router
      os_router:
        state: present
        name: testRouter
        network: public
        cloud: trystack
```
#### Expected Results:

Router successfully created.
#### Actual Results:

Error: 

```
fatal: [localhost]: FAILED! => {
"changed": false, "failed": true, 
"invocation": {"module_args": {"cloud": "trystack", "name": "testRouter", "network": "public", "state": "present"}, "module_name": "os_router"}, 
"msg": "Error creating router testRouter: Policy doesn't allow (rule:create_router and (rule:create_router:external_gateway_info and (rule:create_router:external_gateway_info:network_id and rule:create_router:external_gateway_info:enable_snat))) to be performed."}
```

I did a bit of tinker and I think I found the issue:
1. TryStack policy doesn't allow passing the enable_snat parameter.
2. There is no way to _not_ pass that parameter from the os_router module.

I fixed temporary the issue by allowing None for the `enable_snat` parameter. Looking back, this looks like an acceptable solution. I can provide a quick PR if you agree. ping @emonty @Shrews

**Comment by robynbergeron** [comment]
[module: cloud/openstack/os_router.py]

**Comment by robynbergeron** [comment]
@ibratoev Thanks for submitting this bug report.

@shrews @dguerri @juliakreger @j2sol @rcarrillocruz @emonty -- thanks in advance for taking a look at this bug report and for your ongoing work in maintaining this module.

**Comment by omgjlk** [comment]
This feels like a cloud provider quirk which should be covered by Shade, so that you shouldn't have to know to pass none for `enable_snat`.

https://github.com/openstack-infra/shade is the github mirror of that code.

**Comment by Shrews** [comment]
Thanks for the report.

Out of curiosity, does this happen if enable_snat is True, or False, or both?

If the error occurs with both values, then I think allowing None for the parameter is the best solution here. If the problem happens with only one of the possible values, then the better solution is to fix the problem in shade.

**Comment by Shrews** [comment]
Actually, looking at the neutron API, this value defaults to True. So we can change shade to only explicitly send it if someone sets it to False. I'll toss up a patch to shade for that.

**Comment by Shrews** [comment]
Fix put up: https://review.openstack.org/249931

**Comment by ibratoev** [comment]
`Out of curiosity, does this happen if enable_snat is True, or False, or both?`

Both, I tested with both True and False. TryStack seems to be configured to not allow setting passing this value, even if what you pass is the default...

I found it very easy to setup a test with TryStack - it is a great project. As an idea - a Travis build can be added that runs the Shade tests against a TryStack account. I can help with that ;)
